### PR TITLE
REF (string): de-duplicate ArrowStringArray methods

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
Could we do it the other way around? i.e. have the method in ArrowExtensionArray point to the one defined in ArrowStringArray? 
Personally, I feel it better belongs in string_arrow.py, to keep the string specific functionality in there (the ArrowExtensionArray definition is already huge ..)

**Comment by mroeschke** [comment]
If we're doing this moving of methods away from ArrowExtensionArray, using `pandas.ArrowDtype(pyarrow.string())` will continue to use NA semantics correct?

**Comment by jbrockmendel** [comment]
This REF is behavior-neutral

**Comment by jorisvandenbossche** [comment]
> I'm vaguely planning to move all the relevant stuff up there after the current crop of branches (some still local) are done.

If that's the plan, my preference would be do to it at once, instead of spread over multiple PRs. I know we generally want smaller more granular PRs, but given that this will need to be backported, doing this in one go instead of multiple steps will make that easier.

**Comment by jorisvandenbossche** [comment]
> I think the thing to do is put it in the ArrowStringArrayMixin.

Also a question on the mixin. What's the reason we need the mixin? (compared to keeping things in ArrowStringArray itself) 

Right now ArrowStringArray doesn't actually inherit from the mixin, but uses explicit `_str_method = ArrowStringArrayMixin._str_method` (well, that's actually only done in ArrowStringArrayNumpySemantics, I see, and not in ArrowStringArray, that's another question). 
So using that pattern, we could also keep things in ArrowStringArray, and add the methods to ArrowExtensionArray by doing `_str_method = ArrowStringArray._str_method` in that class. 

Or is the idea that ArrowStringArray(NumpySemantics) eventually actually inherits from ArrowStringArrayMixin?

**Comment by jbrockmendel** [comment]
> Also a question on the mixin. What's the reason we need the mixin? (compared to keeping things in ArrowStringArray itself)

It is not really needed.  ArrowEA subclasses the mixin and ArrowStringArray subclasses ArrowEA.  So all the methods could go directly on ArrowEA, its just that class is already really bulky.  Putting the methods directly on ArrowStringArray and reversing the inheritance would surely be possible but that would be more invasive.

**Comment by jorisvandenbossche** [comment]
Ah, yes, I forgot about ArrowStringArray subclassing ArrowExtensionArray and therefore also subclassing the mixin. 
But the problem with that inheritance is that currently it's not actually used in ArrowStringArray, right? Because ArrowStringArray also subclasses ObjectStringArrayMixin which will get chosen first as fall-back for a method that doesn't exist on ArrowStringArray

That means we should either add the methods explicitly with `_str_method = ArrowStringArrayMixin._str_method`, or let ArrowStringArray subclass ArrowStringArrayMixin directly with the proper order. However, I think the latter does not work? ("Cannot create a consistent method resolution order (MRO)")

**Comment by jbrockmendel** [comment]
> That means we should either add the methods explicitly with _str_method = ArrowStringArrayMixin._str_method

Yes, one of my local de-dup branches does this for the relevant methods.

File: pandas/core/arrays/string_arrow.py [hunk 1]
```diff
@@ -427,47 +407,20 @@ def _str_len(self):
...
-            result = pc.utf8_rtrim(self._pa_array, characters=to_strip)
-        return type(self)(result)
+    _str_match = ArrowExtensionArray._str_match
+    _str_fullmatch = ArrowExtensionArray._str_fullmatch
+    _str_lower = ArrowExtensionArray._str_lower
+    _str_upper = ArrowExtensionArray._str_upper
+    _str_strip = ArrowExtensionArray._str_strip
+    _str_lstrip = ArrowExtensionArray._str_lstrip
+    _str_rstrip = ArrowExtensionArray._str_rstrip
+    _str_removesuffix = ArrowStringArrayMixin._str_removesuffix
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/string_arrow.py:417]: `_str_removesuffix` is also assigned in `ArrowStringArrayNumpySemantics` below. That can then be removed there?

**Comment by jorisvandenbossche** [comment]
If we deduplicate those methods, my preference would be to directly move those methods from ArrowEA to the mixin, and get them from there in ArrowStringArray. Or is there a reason we do not do that here, but use that pattern (shared methods in the mixin instead of in ArrowEA) for other methods?

File: pandas/core/arrays/string_arrow.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/arrays/string_arrow.py:417]: good catch, will update

**Comment by jbrockmendel** [comment]
> my preference would be to directly move those methods from ArrowEA to the mixin, and get them from there in ArrowStringArray.

Sure, will update.  Will wait until after #59615, #59568, #59562 and then rebase on top of those to minimize headaches.

**Review by jorisvandenbossche — COMMENTED** [review]
Looks good! Some small comments

File: pandas/core/arrays/string_arrow.py [hunk 2]
```diff
@@ -335,80 +349,20 @@ def _str_repeat(self, repeats: int | Sequence[int]):
...
-            result = pc.utf8_ltrim(self._pa_array, characters=to_strip)
-        return type(self)(result)
-
-    def _str_rstrip(self, to_strip=None) -> Self:
-        if to_strip is None:
-            result = pc.utf8_rtrim_whitespace(self._pa_array)
-        else:
-            result = pc.utf8_rtrim(self._pa_array, characters=to_strip)
-        return type(self)(result)
+        return ArrowExtensionArray._str_slice(self, start=start, stop=stop, step=step)
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/arrays/string_arrow.py:359]: sure, will update after the slice PR goes in

File: pandas/core/arrays/string_arrow.py [hunk 3]
```diff
@@ -335,80 +349,20 @@ def _str_repeat(self, repeats: int | Sequence[int]):
...
 
     def _str_removeprefix(self, prefix: str):
         if not pa_version_under13p0:
-            starts_with = pc.starts_with(self._pa_array, pattern=prefix)
-            removed = pc.utf8_slice_codeunits(self._pa_array, len(prefix))
-            result = pc.if_else(starts_with, removed, self._pa_array)
-            return type(self)(result)
+            return ArrowExtensionArray._str_removeprefix(self, prefix)
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/arrays/string_arrow.py:363]: sure, will update

**Comment by jorisvandenbossche** [comment]
There are a couple of failures after the latest update

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
Those failures are happening on other PRs as well, I think, so merging.

**Pull request merged**